### PR TITLE
[P0] Send attachment on send

### LINKED ISSUE #5083: Add ability to combine attachments and message text in a single activity
### Is your feature request related to a problem?

Add ability to combine attachments and message text in a single activity

### Describe the suggestion or request in detail

_No response_

### Describe alternatives you have considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #5083.

### Breaking changes

-  `useSendMessage` hook is updated to support sending attachments with a message. To reduce complexity, the `useSendFiles` hook is being deprecated. The hook will be removed on or after 2026-04-03
-  `styleOptions.uploadThumbnailHeight` and `styleOptions.uploadThumbnailWidth` must be a `number` of pixels

### Added

-  Resolves [#5083](https://github.com/microsoft/BotFramework-WebChat/issues/5083). Added `sendAttachmentOn` style option to send attachments and text in a single activity, by [@ms-jb](https://github.com/ms-jb) and [@compulim](https://github.com/compulim)
   -  `useSendMessage` hook is updated to support sending attachments with a message
   -  `useSendBoxAttachments` hook is added to get/set attachments in the send box

## Description

Added a new `styleOption.sendAttachmentOn` that enables attachments and message text to be sent in a single activity. If enabled, uploaded files are retained until the Send button is pressed to allow for message text to be optionally included. Standalone text works as before.

webchat-new-icons

### `useSendFiles` vs. `useSendMessage`

As `useSendMessage` hook now supports sending attachments in the message, the `useSendFiles` hook can be obsoleted.

We choose to add attachments to `useSendMessage` over `useSendFiles` because if we add text message to `useSendFiles`, it will become an uber hook and we would need to deprecate `useSendMessage`. This is undesirable.

### `sendBoxAttachments` hook

New `useSendBoxAttachments` to get/set the attachments in the send box before it is being sent. This works nicely with speech recognition.

### Simplifying attachments

Some philosophies we applied to simply the SDK design:

- Userland attachment is `{ blob: Blob | File, thumbnailURL?: URL }`, as minimal/intuitive as possible
   - Also extensible, say, `{ iconURL?: URL }`
- `botframework-webchat-component` will help making thumbnails when possible
   - `useSendMessage` will accept userland attachment. When the hook is called, thumbnails will be generated on-the-fly
   - `useSendFiles` will remain the same and accept `Blob | File`. When the hook is called, thumbnails will be generated on-the-fly
   - `useSendBoxAttachments` will accept userland attachment. However, it will not automatically generate thumbnails

## Specific Changes

- Hooks
   - Added `useSendBoxAttachments` hook to get/set attachments in the send box
   - Modified `useSendMessage` hook to add `attachments`
   - Deprecated `useSendFiles` hook
- UI
   - Updated `<UploadButton>` to show checkmark when files are attached
   - Added `styleOptions.sendAttachmentOn`
      - `"attach"` will send the message with attachment when files are attached (legacy)
      - `"send"` will send the message with attachment when the send button is tapped (default)

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] I have updated documentation

## Review Checklist

> This section is for contributors to review your work.

-  [x] Accessibility reviewed (tab order, content readability, alt text, color contrast)
-  [x] Browser and platform compatibilities reviewed
-  [x] CSS styles reviewed (minimal rules, no `z-index`)
-  [x] Documents reviewed (docs, samples, live demo)
-  [x] Internationalization reviewed (strings, unit formatting)
-  [x] `package.json` and `package-lock.json` reviewed
-  [x] Security reviewed (no data URIs, check for nonce leak)
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

File: packages/api/src/hooks/internal/WebChatAPIContext.ts [hunk 1]
```diff
@@ -1,27 +1,35 @@
...
+  WebChatPostActivityAttachment
+} from 'botframework-webchat-core';
 
 type WebChatAPIContext = {
   activityRenderer?: LegacyActivityRenderer;
   activityStatusRenderer: RenderActivityStatus;
   attachmentForScreenReaderRenderer?: AttachmentForScreenReaderComponentFactory;
   attachmentRenderer?: RenderAttachment;
-  avatarRenderer?: AvatarComponentFactory;
+  avatarRenderer: AvatarComponentFactory;
```
Comment (compulim) [inline packages/api/src/hooks/internal/WebChatAPIContext.ts:32]: Fixing typing error.

File: packages/api/src/hooks/internal/WebChatAPIContext.ts [hunk 2]
```diff
@@ -45,17 +53,21 @@ type WebChatAPIContext = {
     linkOptions: { externalLinkAlt: string }
   ) => string;
   scrollToEndButtonRenderer?: ScrollToEndButtonComponentFactory;
-  selectVoice?: (voices: (typeof window.SpeechSynthesisVoice)[], activity: WebChatActivity) => void;
+  selectVoice?: (
+    voices: (typeof window.SpeechSynthesisVoice)[],
+    activity: WebChatActivity
+  ) => typeof window.SpeechSynthesisVoice;
   sendEvent?: (name: string, value: any) => void;
-  sendFiles?: (files: File[]) => void;
```
Comment (compulim) [inline packages/api/src/hooks/internal/WebChatAPIContext.ts:50]: Previously, `(files: File[]) => void` was wrong, it should be `(files: WebChatAttachment[]) => void`.

File: packages/api/src/hooks/internal/useWebChatAPIContext.ts
Comment (compulim) [inline packages/api/src/hooks/internal/useWebChatAPIContext.ts:1]: Renaming to `.ts` will enable typechecking down the stream.

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "botframework-webchat-root",
-  "version": "4.16.1-0",
+  "version": "4.17.0-0",
```
Comment (compulim) [inline package.json:3]: Bumping minor (a.k.a. major) as we are deprecating `useSendFiles` hook.

File: packages/api/src/hooks/Composer.tsx [hunk 4]
```diff
@@ -198,7 +195,13 @@ type ComposerCoreProps = Readonly<{
...
   directLine: DirectLineJSBotConnection;
   disabled?: boolean;
-  downscaleImageToDataURL?: (blob: Blob, maxWidth: number, maxHeight: number, type: string, quality: number) => string;
+  downscaleImageToDataURL?: (
+    blob: Blob,
+    maxWidth: number,
+    maxHeight: number,
+    type: string,
+    quality: number
+  ) => Promise<URL>;
```
Comment (compulim) [inline packages/api/src/hooks/Composer.tsx:204]: Fixing typing mistakes.

File: packages/api/src/hooks/useTrackTiming.ts [hunk 5]
```diff
@@ -2,30 +2,31 @@ import { useCallback } from 'react';
...
-import useTrackException from './useTrackException';
 import useWebChatAPIContext from './internal/useWebChatAPIContext';
+import usePonyfill from './usePonyfill';
+import useTrackException from './useTrackException';
 
-export default function useTrackTiming<T>(): (
-  name: string,
-  functionOrPromise: (() => T) | Promise<T>
-) => Promise<T | void> {
+export default function useTrackTiming<T>(): (name: string, functionOrPromise: (() => T) | Promise<T>) => Promise<T> {
```
Comment (compulim) [inline packages/api/src/hooks/useTrackTiming.ts:10]: Fix typing mistakes.

File: packages/api/src/hooks/internal/WebChatAPIContext.ts [hunk 6]
```diff
@@ -1,33 +1,46 @@
...
   disabled?: boolean;
   dismissNotification?: (id: string) => void;
-  downscaleImageToDataURL?: (blob: Blob, maxWidth: number, maxHeight: number, type: string, quality: number) => string;
+  downscaleImageToDataURL?: (
+    blob: Blob,
+    maxWidth: number,
+    maxHeight: number,
+    type: string,
+    quality: number
+  ) => Promise<URL>;
```
Comment (compulim) [inline packages/api/src/hooks/internal/WebChatAPIContext.ts:43]: Fix typing mistakes.

File: packages/component/src/Utils/downscaleImageToDataURL/blobToArrayBuffer.js
Comment (compulim) [inline packages/component/src/Utils/downscaleImageToDataURL/blobToArrayBuffer.js:1]: No longer needed as we have `Blob.prototype.arrayBuffer(): Promise<ArrayBuffer>`.

File: __tests__/hooks/useSendFiles.js [hunk 7]
```diff
@@ -27,7 +27,7 @@ test('calling sendFile should send files', async () => {
     sendFiles([blob1, blob2]);
   });
 
-  await driver.wait(minNumActivitiesShown(2), timeouts.directLine);
+  await driver.wait(minNumActivitiesShown(3), timeouts.directLine);
```
Comment (compulim) [inline __tests__/hooks/useSendFiles.js:30]: Fix tests.

**Review by cwhitten — APPROVED** [review]

**Pull request merged**